pkgs/top-level/all-packages.nix

```diff
@@ -9210,24 +9210,66 @@ in
     git = gitMinimal;
   };
 
+  # Build summary for this flavor
+  # - Build Octave Qt GUI:                  no (missing:QtCore QtGui QtNetwork QtHelp QtXml)
+  # - JIT compiler for loops:               no
+  # - Build Java interface:                 yes
+  # - Build static libraries:               no
+  # - Build shared libraries:               yes
+  # - Dynamic Linking API:                  dlopen
+  # - Include support for GNU readline:     yes
+  # - 64-bit array dims and indexing:       yes
+  # - 64-bit BLAS array dims and indexing:  yes
+  # - OpenMP SMP multithreading:            yes
+  # - Truncate intermediate FP results:     yes
+  # - Build cross tools:                    no
+  # - Build docs:                           yes
   octave = callPackage ../development/interpreters/octave {
-    qt = null;
-    qscintilla = null;
-    ghostscript = null;
-    graphicsmagick = null;
-    llvm = null;
-    hdf5 = null;
-    glpk = null;
-    suitesparse = null;
-    jdk = null;
+    python = python3;
+    sundials = sundials_2;
```

Comment:
Same here, if we explicitly need `sundials_2`, we should make that clear in the function signature.

Reply:
That's just my taste on how to use specific versions when defining packages. The idea is that in the future, when sundials 5 support will finally land in octave, the diff will be cleaner on octave's `default.nix`.

A similar idea goes for `python` dependencies - when in the future `python3` will be the default, we'll need to update mostly `all-packages.nix` and not all the different files in which `python3` is used while it could be replaced with just `python`. It's easier to spot packages that don't support a latest version of a software and be aware of that when this version specification is done in `all-packages.nix`.

That's just my opinion so I wouldn't mind changing this bits here. I hope you are satisfied with the more significant changes I made.

Reply:
Yes, I really appreciate the other changes.

However, I think it'd be more clear if we'd explicitly request `sundials_2` in the function header inside `octave/default.nix` instead of overriding things in `all-packages.nix`. Looking at the upstream discussion, it seems likely we'll need to explicitly pull in a specific sundials version compatible with the current octave version anyways.

Python 3 is a bit different. I think flipping `pkgs.python` to point to `pkgs.python3` nixpkgs-wide would cause much more breakage, which is why it hasn't happened yet.

However, people still want to get rid of python2 in their closure, so they either request python3 from the function header, or by passing python3 as python into the package. To reduce complexity, just moving it into the function header might be fine aswell - it'll probably still be quite some time till there's a `python4` :-)